### DISCUSSION TITLE
Unable to install app on physical iOS device

### DISCUSSION THREAD

**Discussion by Gys** [fyne-io/fyne#4939] [General]
I have a Fyne app that I need to test on a physical iOS device. It is built on an MacBook with xcode, I have a paid developer account, the device (UDID) is registered with xcode and online in my Apple developer account. I have built and installed iOS apps before, but those were directly from xcode.

If I try to install a Fyne app on a physical iPhone, xcode says: `This app cannot be installed because its integrity could not be verified.`. And in the details: `No code signature found.`

I have read https://docs.fyne.io/started/mobile and https://github.com/fyne-io/fyne-cross.

I did:

`fyne package -os ios -appID com.example.myapp -icon myicon.png`

On https://docs.fyne.io/started/mobile it says: For iOS to install on device open Xcode and choose the “Devices and Simulators” menu item within the “Window” menu. Then find your phone and drag the myapp.app icon onto your app list. 

But if I do that the error message above appears.

I can find several potential solutions for this problem, but it all involves opening a project in xcode and click around there. However, in the case of Fyne this is not an option? 

I cannot find any reference otherwise to this problem in any Fyne related issues or discussions.

**Comment by andydotxyz** [comment]
If you have the certification and provisioning profile from Apple then you're just missing the "-profile" and "-certificate" parameters to the build. The default uses anonymous developer credentials which is not what you need from what is described above.

**Reply by Gys** [reply]
> the certification and provisioning profile

You make it sound if there is just one of each, but it seems to me there are several options for several versions / types. To narrow it down: can I create those inside xcode? Or do I create those online in the developer account?

**Reply by andydotxyz** [reply]
You'll want to go online to their certificate / profile management pages and create/download the appropriate files. My apologies for brevity, but you said you had done this before so do not want to appear patronising.

**Reply by Gys** [reply]
Sorry, I merely mentioned that to avoid comments that I need xcode, developer account, etc. I have very limited knowledge of the cryptographic stuff. Inside xcode it mostly just works because xcode can manage certificates and profiles. It is also easy to find online guidance on what to click where inside xcode. But Fyne manages the project so its not clear how to do 'project things' without access to the project.

Which brings me to something else:

I have a proof of concept app accessing the camera. However, for accessing hardware features like the camera, some settings in the project are needed. Settings saved in the Info.plist. See below. 

I think fyne / fyne-cross generates a new Info.plist on the fly for each build? And there is no option to add / modify elements of it?

[Configure access alerts](https://developer.apple.com/documentation/avfoundation/capture_setup/requesting_authorization_to_capture_and_save_media#2962313)

`The system requires that your app provides static messages to display to the user when the system asks for permission to access the camera or microphone...`.

`For each key, provide a message that explains to the user why your app needs to capture media so the user can feel confident granting permission to your app.`

**Reply by andydotxyz** [reply]
> But Fyne manages the project so its not clear how to do 'project things' without access to the project.

Fyne does not manage the certification in a way different to Xcode - you will need the provisioning profile and so forth set up for your iOS target app - any Apple documentation or forum posts will step through. The process for a Fyne app should be the same as a native iOS app except the package/compile step.

**Reply by andydotxyz** [reply]
> I think fyne / fyne-cross generates a new Info.plist on the fly for each build? And there is no option to add / modify elements of it?

For this it may be a feature request time - our packaging will need a feature to enable certain extra access as the same is needed for Android as well.

**Reply by Gys** [reply]
The Apple docs say:

`Important

Your app needs to contain the appropriate key in its Info.plist file, and the appropriate entitlement enabled in macOS, before it requests authorization or attempts to use a capture device. Otherwise, the system terminates your app.`

And indeed, as soon as I click the button to access the camera the app terminates. At least it confirms I implemented the first step successfully...

Edit: although, it says MacOS here, so not sure what that means on iOS. It does terminate when I click the button.

**Reply by andydotxyz** [reply]
Indeed that is probably to be expected without the additional info, which is why I suggested opening a request that these extra features could be added to packaged metadata. 

This discussion isn't the right place to track the changes required to support additional permissions or metadata.

**Comment by Gys** [comment]
Answering this here for future reference.

Asking permission for using the camera on iOS is required, otherwise the app simply crashes.

Asking is simple by adding these two lines to the Info.plist file:

    <key>NSCameraUsageDescription</key>
    <string>Camera access is required for this app</string>

Where the string should be more specific in describing the requirement. The app store acceptance can be picky about a too general question.

The info.plist file is generated by Fyne and therefore a hack is needed (at this moment). You need to edit the file `cmd/fyne/internal/mobile/build_iosapp.go` inside the Fyne package and then rebuild the `fyne` command before using `fyne` to build the ios version.
